src/sources/util/body_decoding.rs

```diff
@@ -41,37 +46,48 @@ fn body_to_lines(buf: Bytes) -> impl Iterator<Item = Result<Bytes, ErrorMessage>
 }
 
 #[cfg(any(feature = "sources-http", feature = "sources-datadog"))]
-pub fn decode_body(body: Bytes, enc: Encoding) -> Result<Vec<Event>, ErrorMessage> {
+pub fn decode_body(
+    body: Bytes,
+    enc: Encoding,
+    batch: Arc<BatchNotifier>,
+) -> Result<Vec<Event>, ErrorMessage> {
     match enc {
         Encoding::Text => body_to_lines(body)
-            .map(|r| Ok(Event::from(r?)))
+            .map(|r| {
+                Ok(LogEvent::from(r?)
+                    .with_batch_notifier(Arc::clone(&batch))
+                    .into())
+            })
             .collect::<Result<_, _>>(),
         Encoding::Ndjson => body_to_lines(body)
             .map(|j| {
                 let parsed_json = serde_json::from_slice(&j?)
                     .map_err(|error| json_error(format!("Error parsing Ndjson: {:?}", error)))?;
-                json_parse_object(parsed_json)
+                json_parse_object(parsed_json, &batch)
+                    .map(|log| log.with_batch_notifier(Arc::clone(&batch)).into())
             })
             .collect::<Result<_, _>>(),
         Encoding::Json => {
             let parsed_json = serde_json::from_slice(&body)
                 .map_err(|error| json_error(format!("Error parsing Json: {:?}", error)))?;
-            json_parse_array_of_object(parsed_json)
+            json_parse_array_of_object(parsed_json, batch)
         }
     }
 }
 
 #[cfg(any(feature = "sources-http", feature = "sources-datadog"))]
-fn json_parse_object(value: JsonValue) -> Result<Event, ErrorMessage> {
-    let mut event = Event::new_empty_log();
-    let log = event.as_mut_log();
-    log.insert(log_schema().timestamp_key(), Utc::now()); // Add timestamp
+fn json_parse_object(
+    value: JsonValue,
+    batch: &Arc<BatchNotifier>,
+) -> Result<LogEvent, ErrorMessage> {
     match value {
         JsonValue::Object(map) => {
+            let mut log = LogEvent::default();
+            log.insert(log_schema().timestamp_key(), Utc::now()); // Add timestamp
             for (k, v) in map {
                 log.insert_flat(k, v);
             }
-            Ok(event)
+            Ok(log.with_batch_notifier(Arc::clone(batch)))
```

Comment:
It seems a little unfortunate to plumb this all the way down into what's otherwise a pretty simple parsing function. Would it be reasonable to maybe pull this up somewhere closer to where we send the events through the `Pipeline`?

Reply:
I'll see what I can do to lift it upwards in the stack.

Reply:
Thanks for the suggestion! Other than some work in the event metadata, this resulted in a much more generic approach to handling the notifier.